### PR TITLE
Add support for App Store Connect API Key

### PR DESCRIPTION
**Pull request by knox**
This adds support for App Store Connect API Keys

The use of API Keys as authentication method does not only enables deliver to use modern, documented API but it also overcomes problems with 2FA and other legacy authentication methods.

This implementation started as a copy-and-paste from [steps-deploy-to-itunesconnect-application-loader](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-application-loader) but had to be tweaked and extended until it finally worked.

Please note that as per [documentation](https://docs.fastlane.tools/actions/deliver/#parameters) fastlane deliver accepts an `--api_key` parameter to pass in API Key information but that is reported to [not work](https://github.com/fastlane/fastlane/issues/17528) and said to even being [not supported](https://github.com/fastlane/fastlane/issues/17528#issuecomment-723730863) at all. So API Key information must be written into a [special crafted](https://docs.fastlane.tools/app-store-connect-api/#using-fastlane-api-key-json-file) JSON File and it's path passed to deliver with the `--api_key_path` parameter instead.

### CODE REVIEW

**Comment by godrei** [comment]
Hi @knox 
thanks for the contribution, i started to review the changes.

The CI failed by Go static check issues:

```
Configs:
- packages: github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver
github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/devportalservice

Installing golint
$ go "get" "-u" "golang.org/x/lint/golint"

Running golint...
$ golint "-set_exit_status" "github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver"
golint failed
/Users/vagrant/go/src/github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/main.go:350:6: exported type ApiKey should have comment or be unexported
/Users/vagrant/go/src/github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/main.go:350:6: type ApiKey should be APIKey
Found 2 lint suggestions; failing.
golint failed: exit status 1
```

[I separated the `go-tests` workflow](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/pull/88/commits/3295c65e455a2632f8265f6c4e61dc039cb35140), which runs all of our Go checks (it worth to run locally before pushing changes) and [fixed the issues](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/pull/88/commits/efd1e42deea9b4f7c1ee4121e5df525f5aaa10f7).

**Comment by knox** [comment]
Thanks @godrei for having a look at this.

To be honest these are almost my very first lines of golang and i used a simple text editor to author them. I will try to figure out why tests are failing and hopefully fix it.

**Comment by knox** [comment]
@godrei Unfortunately i was not able to run `go-tests` on my macbook so i have filed https://github.com/bitrise-steplib/steps-golint/issues/6 and https://github.com/bitrise-steplib/steps-go-list/issues/7 

Maybe i should just stick to things i know about 🤷

**Comment by godrei** [comment]
@knox I'm working on the CI setup, generating API key, and planning a new test workflow for these changes.

Originally I was testing this PR: https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/pull/77 on this branch.

I could not make the login successful without the normal password, even if an app-specific password is provided.
This is why the latest build is failing (build log at the end).

deliver [uses iTMSTransporter for binary, and spaceship for metadata upload](https://docs.fastlane.tools/actions/deliver/#how-does-this-thing-even-work-is-magic-involved).
iTMSTransporter can authenticate using an API key or apple id + password/app-specific password.
[spaceship by default uses API key auth](https://github.com/fastlane/fastlane/blob/master/spaceship/docs/AppStoreConnect.md#usage), but in the case of apple id auth it switches to the old tunes client, wich requires an apple id + password + TFA session, in the case of two-factor enabled accounts.

I will revert [these change](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/pull/88/commits/dc9fecc8e4e177f4c0aba43fe337e7b6347ba0ac#diff-61d1ee831cb1df242161a7d23171914a1d5d628c7659b52db957d80a9394bd68R82) and get back here with the build result.

```
INFO[21:31:40]  * [OK] Step dependency (go) installed, available.
Config:
- IpaPath: /Users/vagrant/deploy/Bitrise-iTunesConnectBetaTest.ipa
- PkgPath:
- ItunesConnectUser: [REDACTED]
- Password:
- AppPassword: *****
- APIKeyPath:
- APIIssuer:
- AppID: [REDACTED]
- BundleID:
- SubmitForReview: no
- SkipMetadata: yes
- SkipScreenshots: no
- SkipAppVersionUpdate: no
- TeamID: [REDACTED]
- TeamName:
- Platform: ios
- Options:
- GemfilePath: ./Gemfile
- FastlaneVersion: latest
- ITMSParameters: -t DAV
- VerboseLog: true

Connected session-based Apple Developer Portal Account found

Setup
fastlane version defined: latest, installing...

$ gem "install" "fastlane" "--no-document" "--prerelease"

$ rbenv "rehash"

$ fastlane "-v"
fastlane installation at path:
/Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/bin/fastlane
-----------------------------
fastlane 2.171.0
Setup took 11.524468 seconds to complete

Deploy
**Note:** if your password
contains special characters
and you experience problems, please
consider changing your password
to something with only
alphanumeric characters.

**Be advised**
that this step uses a well maintained, open source tool which
uses *undocumented and unsupported APIs* (because the current
iTunes Connect platform does not have a documented and supported API)
to perform the deployment.
This means that when the API changes
**this step might fail until the tool is updated**.


$ fastlane "deliver" "--username" "[REDACTED]" "--app" "[REDACTED]" "--team_id" "[REDACTED]" "--ipa" "/var/folders/6q/wgy6jtp12w5gzgm9lzcglpqw0000gn/T/ipaOrPkg958433879/tmp.ipa" "--skip_metadata" "--force" "--platform" "ios"

[21:31:56]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
[21:31:57]: Sending anonymous analytics information
[21:31:57]: Learn more at https://docs.fastlane.tools/#metrics
[21:31:57]: No personal or sensitive data is sent.
[21:31:57]: You can disable this by adding `opt_out_usage` at the top of your Fastfile
[21:31:57]: Login to App Store Connect ([REDACTED])
-------------------------------------------------------------------------------------
Please provide your Apple Developer Program account credentials
The login information you enter will be stored in your macOS Keychain
You can also pass the password using the `FASTLANE_PASSWORD` environment variable
See more information about it on GitHub: https://github.com/fastlane/fastlane/tree/master/credentials_manager
-------------------------------------------------------------------------------------

Looking for related GitHub issues on fastlane/fastlane...

Found no similar issues. To create a new issue, please visit:
https://github.com/fastlane/fastlane/issues/new
Run `fastlane env` to append the fastlane environment to your issue
/Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/credentials_manager/lib/credentials_manager/account_manager.rb:135:in `ask_for_login': [!] Missing password for user [REDACTED], and running in non-interactive shell (RuntimeError)
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/credentials_manager/lib/credentials_manager/account_manager.rb:56:in `password'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/client.rb:380:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/client.rb:352:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/tunes/spaceship.rb:24:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/client.rb:56:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/spaceship.rb:85:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/runner.rb:36:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/runner.rb:22:in `initialize'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:81:in `new'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:81:in `block (2 levels) in run'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:186:in `run'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:17:in `start'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/cli_tools_distributor.rb:111:in `take_off'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/bin/fastlane:23:in `<top (required)>'
	from /Users/vagrant/.rbenv/versions/2.6.5/bin/fastlane:23:in `load'
	from /Users/vagrant/.rbenv/versions/2.6.5/bin/fastlane:23:in `<main>'
/Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/credentials_manager/lib/credentials_manager/account_manager.rb:135:in `ask_for_login': Missing password for user [REDACTED], and running in non-interactive shell (RuntimeError)
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/credentials_manager/lib/credentials_manager/account_manager.rb:56:in `password'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/client.rb:380:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/client.rb:352:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/tunes/spaceship.rb:24:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/client.rb:56:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/spaceship.rb:85:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/runner.rb:36:in `login'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/runner.rb:22:in `initialize'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:81:in `new'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:81:in `block (2 levels) in run'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:186:in `run'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/deliver/lib/deliver/commands_generator.rb:17:in `start'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/cli_tools_distributor.rb:111:in `take_off'
	from /Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/bin/fastlane:23:in `<top (required)>'
	from /Users/vagrant/.rbenv/versions/2.6.5/bin/fastlane:23:in `load'
	from /Users/vagrant/.rbenv/versions/2.6.5/bin/fastlane:23:in `<main>'
Deploy failed, error: exit status 1
```

**Comment by godrei** [comment]
If we [turn on the password](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/commit/615a9ce36d0c5611fc91d1ef805eaf6820e8c60e) the CI build succeed.

I suggest restoring the original CI first, to make sure these code changes will not break the apple id + password auth. After that, we should figure out how to test both the old and new authentication.

```
INFO[22:06:48]  * [OK] Step dependency (go) installed, available.
Config:
- IpaPath: /Users/vagrant/deploy/Bitrise-iTunesConnectBetaTest.ipa
- PkgPath:
- ItunesConnectUser: [REDACTED]
- Password: *****
- AppPassword: *****
- APIKeyPath:
- APIIssuer:
- AppID: [REDACTED]
- BundleID:
- SubmitForReview: no
- SkipMetadata: yes
- SkipScreenshots: no
- SkipAppVersionUpdate: no
- TeamID: [REDACTED]
- TeamName:
- Platform: ios
- Options:
- GemfilePath: ./Gemfile
- FastlaneVersion: latest
- ITMSParameters: -t DAV
- VerboseLog: true

Connected session-based Apple Developer Portal Account found

Setup
fastlane version defined: latest, installing...

$ gem "install" "fastlane" "--no-document" "--prerelease"

$ rbenv "rehash"

$ fastlane "-v"
fastlane installation at path:
/Users/vagrant/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/bin/fastlane
-----------------------------
fastlane 2.171.0
Setup took 12.139496 seconds to complete

Deploy
**Note:** if your password
contains special characters
and you experience problems, please
consider changing your password
to something with only
alphanumeric characters.

**Be advised**
that this step uses a well maintained, open source tool which
uses *undocumented and unsupported APIs* (because the current
iTunes Connect platform does not have a documented and supported API)
to perform the deployment.
This means that when the API changes
**this step might fail until the tool is updated**.


$ fastlane "deliver" "--username" "[REDACTED]" "--app" "[REDACTED]" "--team_id" "[REDACTED]" "--ipa" "/var/folders/6q/wgy6jtp12w5gzgm9lzcglpqw0000gn/T/ipaOrPkg682151402/tmp.ipa" "--skip_metadata" "--force" "--platform" "ios"

[22:07:05]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
[22:07:06]: Sending anonymous analytics information
[22:07:06]: Learn more at https://docs.fastlane.tools/#metrics
[22:07:06]: No personal or sensitive data is sent.
[22:07:06]: You can disable this by adding `opt_out_usage` at the top of your Fastfile
[22:07:06]: Login to App Store Connect ([REDACTED])
[22:07:07]: Login successful
```

**Comment by knox** [comment]
> I could not make the login successful without the normal password, even if an app-specific password is provided.

I'm pretty sure i 
[got it working](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/issues/85#issuecomment-732404421) but we should not waste time with it. Application specific passwords are dead and API Keys is the way to go.

> I suggest restoring the original CI first, to make sure these code changes will not break the apple id + password auth. 

I will take care of this and push it. We still have to make the password optional so that just API Key path and issuer becomes a valid configuration.

> After that, we should figure out how to test both the old and new authentication.

For this some more assistance of yours will be required.

**Comment by knox** [comment]
@godrei i have added an experimental test case for API Key authentication. It's disabled for now because the testing is hard-tied to the bitrise beta app and only you guys can codesign it.

> error: No profiles for 'com.bitrise.Bitrise-iTunesConnectBetaTest' were found

So tests fail for everybody else. Even the pre-existing test did.

File: bitrise.yml [hunk 1]
```diff
@@ -64,11 +64,11 @@ workflows:
...
   test-api-key-auth:
+    title: Test API Key authentication
     before_run:
       - prepare-test
     steps:
     - path::./:
-        title: Step Test password authentication
         inputs:
           - api_key_path: $ITUNES_CONNECT_API_KEY_ISSUER
           - api_issuer: $ITUNES_CONNECT_API_KEY_ISSUER
```
Comment (godrei) [inline bitrise.yml:74]: should be: `ITUNES_CONNECT_API_KEY_PATH `

**Comment by godrei** [comment]
@knox i updated the CI setup:
- API key accessible via `ITUNES_CONNECT_API_KEY_PATH` env
- issuer id available in `ITUNES_CONNECT_API_KEY_ISSUER` env

you can turn on the `test-api-key-auth` workflow.

**Comment by knox** [comment]
> i updated the CI setup
> you can turn on the test-api-key-auth workflow.

It fails but i can't access the build log. Could you please have a look?

**Comment by godrei** [comment]
It seems we are close:

```
15:07:49]: Making sure the latest version on App Store Connect matches '4.0'...
[15:07:50]: '4.0' is the latest version on App Store Connect
[15:07:50]: Will begin uploading snapshots for '4.0' on App Store Connect
[15:07:50]: Starting with the upload of screenshots...
[15:07:50]: Previous uploaded. Skipping './screenshots/en-US/1_ipadPro_1.2048 x 2732.jpeg'...
[15:07:50]: Previous uploaded. Skipping './screenshots/en-US/2_ipadPro_2.2732 x 2048.jpeg'...
[15:07:50]: Previous uploaded. Skipping './screenshots/en-US/1_iphone6Plus_1.1242 x 2208.jpeg'...
[15:07:50]: Previous uploaded. Skipping './screenshots/en-US/2_iphone6Plus_2.2208 x 1242.jpeg'...
[15:07:50]: Waiting for all the screenshots processed...
[15:07:51]: Successfully uploaded all screenshots
[15:07:51]: Sorting screenshots uploaded...
[15:07:51]: Successfully uploaded screenshots to App Store Connect
[15:07:51]: Uploading binary to App Store Connect
[15:07:51]: Going to upload updated app to App Store Connect
[15:07:51]: This might take a few minutes. Please don't interrupt the script.
[15:07:51]: Apple recommends you don’t specify the -t transport and instead allow Transporter to use automatic transport discovery to determine the best transport mode for your packages. For more information, please read Apple's Transporter User Guide 2.1: https://help.apple.com/itc/transporteruserguide/#/apdATD1E1288-D1E1A1303-D1E1288A1126
[15:07:51]: Apple recommends you don’t specify the -t transport and instead allow Transporter to use automatic transport discovery to determine the best transport mode for your packages. For more information, please read Apple's Transporter User Guide 2.1: https://help.apple.com/itc/transporteruserguide/#/apdATD1E1288-D1E1A1303-D1E1288A1126
[15:08:17]: iTunes Transporter successfully finished its job
[15:08:17]: --------------------------------------------------------------------
[15:08:17]: Successfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.
[15:08:17]: --------------------------------------------------------------------
[15:08:17]: Finished the upload to App Store Connect
[15:08:17]: Running precheck before submitting to review, if you'd like to disable this check you can set run_precheck_before_submit to false
[15:08:17]: Running precheck 👮‍♀️ 👮
[!] Precheck cannot check In-app purchases with the App Store Connect API Key (yet). Exclude In-app purchases from precheck or use Apple ID login
Deploy failed, error: exit status 1
```

**Comment by godrei** [comment]
```
[15:27:38]: [Transporter Error Output]: ERROR ITMS-90189: "Redundant Binary Upload. You've already uploaded a build with build number '854' for version number '4.0'. Make sure you increment the build string before you upload your app to App Store Connect. Learn more in Xcode Help (http://help.apple.com/xcode/mac/current/#/devba7f53ad4)."
[15:27:38]: Transporter transfer failed.
[15:27:38]: 
[15:27:38]: ERROR ITMS-90189: "Redundant Binary Upload. You've already uploaded a build with build number '854' for version number '4.0'. Make sure you increment the build string before you upload your app to App Store Connect. Learn more in Xcode Help (http://help.apple.com/xcode/mac/current/#/devba7f53ad4)."
```

Might we should set the build number this way: `($BITRISE_BUILD_NUMBER * 10)` in `test-password-auth` workflow and `($BITRISE_BUILD_NUMBER * 10) + 1` in `test-api-key-auth` workflow.

WDYT?

**Comment by knox** [comment]
🎉

**Review by godrei — APPROVED** [review]

**Comment by godrei** [comment]
🚀 Congrats @knox and huge thanks for the contribution!

**Pull request merged**